service/controller/v18/adapter/guest_subnets.go

```diff
@@ -16,14 +20,44 @@ type GuestSubnetsAdapter struct {
 }
 
 func (s *GuestSubnetsAdapter) Adapt(cfg Config) error {
+	publicSubnet, privateSubnet, err := allocatePublicAndPrivateSubnets(cfg)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
 	s.PublicSubnetAZ = key.AvailabilityZone(cfg.CustomObject)
-	s.PublicSubnetCIDR = cfg.CustomObject.Spec.AWS.VPC.PublicSubnetCIDR
+	s.PublicSubnetCIDR = publicSubnet.String()
 	s.PublicSubnetName = key.SubnetName(cfg.CustomObject, suffixPublic)
 	s.PublicSubnetMapPublicIPOnLaunch = false
 	s.PrivateSubnetAZ = key.AvailabilityZone(cfg.CustomObject)
-	s.PrivateSubnetCIDR = key.PrivateSubnetCIDR(cfg.CustomObject)
+	s.PrivateSubnetCIDR = privateSubnet.String()
 	s.PrivateSubnetName = key.SubnetName(cfg.CustomObject, suffixPrivate)
 	s.PrivateSubnetMapPublicIPOnLaunch = false
 
 	return nil
 }
+
+func allocatePublicAndPrivateSubnets(cfg Config) (net.IPNet, net.IPNet, error) {
+	_, subnet, err := net.ParseCIDR(key.ClusterNetworkCIDR(cfg.CustomObject))
+	if err != nil {
+		return net.IPNet{}, net.IPNet{}, microerror.Mask(err)
+	}
+
+	privateSubnetMask := net.CIDRMask(cfg.PrivateSubnetMaskBits, 32)
+	publicSubnetMask := net.CIDRMask(cfg.PublicSubnetMaskBits, 32)
+
+	var reservedSubnets []net.IPNet
+	privateSubnet, err := ipam.Free(*subnet, privateSubnetMask, reservedSubnets)
+	if err != nil {
+		return net.IPNet{}, net.IPNet{}, microerror.Mask(err)
+	}
+
+	reservedSubnets = append(reservedSubnets, privateSubnet)
+
+	publicSubnet, err := ipam.Free(*subnet, publicSubnetMask, reservedSubnets)
+	if err != nil {
+		return net.IPNet{}, net.IPNet{}, microerror.Mask(err)
+	}
```

Comment:
Yep, but so far size of private/public subnets has been configurable (I don't know if it actually makes sense) so those can differ in size.

Reply:
I don't think it does. Half is the biggest subnet you can assign. What is the point to configure it to be smaller?

Reply:
You could do this:
* **Tenant subnet:** `10.100.0.0/22`
* **Public subnet:** `10.100.0.0/24`
* **Private subnet:** `10.100.2.0/23`

But I don't think there's any point in this because then you'd waste `10.100.1.0/24`. I also don't know where this comes from so I think it makes sense to find out. I'll ask people.

Reply:
You can but why not

* **Tenant subnet:** `10.100.0.0/22`
* **Public subnet:** `10.100.1.0/23`
* **Private subnet:** `10.100.2.0/23`